### PR TITLE
SM/MM: fix possible sending in out-of-order.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1156/ for details.

File: src/uct/sm/mm/mm_ep.c [hunk 1]
```diff
@@ -43,6 +43,8 @@ static UCS_CLASS_INIT_FUNC(uct_mm_ep_t, uct_iface_t *tl_iface,
     self->mapped_desc.mmid   = remote_iface_addr->id;
     uct_mm_set_fifo_ptrs(self->mapped_desc.address, &self->fifo_ctl, &self->fifo);
 
+    self->cached_tail = self->fifo_ctl->tail;
```
Comment (miked-mellanox) [inline src/uct/sm/mm/mm_ep.c]: check

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/568/ for details (Mellanox internal link).

**Comment by alinask** [comment]
@miked-mellanox @yosefe please review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1158/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/569/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
:+1:

File: src/uct/sm/mm/mm_ep.c [hunk 2]
```diff
@@ -194,9 +196,20 @@ uct_mm_ep_am_common_send(const unsigned is_short, uct_mm_ep_t *ep, uct_mm_iface_
     UCT_CHECK_AM_ID(am_id);
 
     head = ep->fifo_ctl->head;
-    /* check if there is room in the remote process's receive fifo to write */
-    if (ucs_unlikely((head - ep->fifo_ctl->tail) >= iface->config.fifo_size)) {
-        return UCS_ERR_NO_RESOURCE;
+    /* check if there is room in the remote process's receive FIFO to write */
+    if (!UCT_MM_EP_IS_ABLE_TO_SEND(head, ep->cached_tail, iface->config.fifo_size)) {
+        if (ep->arb_group.tail != NULL) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: pls add inline function to arbiter such as `ucs_arbiter_group_is_empty()` instead of accessing it directly

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1159/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/570/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
autobot build

**Comment by yosefe** [comment]
:+1:

**Comment by yosefe** [comment]
@shamisp looks like buildbot is not running - can we merge?

**Comment by shamisp** [comment]
let's wait for our testing results please

File: src/uct/sm/mm/mm_ep.c [hunk 3]
```diff
@@ -194,9 +196,20 @@ uct_mm_ep_am_common_send(const unsigned is_short, uct_mm_ep_t *ep, uct_mm_iface_
     UCT_CHECK_AM_ID(am_id);
 
     head = ep->fifo_ctl->head;
-    /* check if there is room in the remote process's receive fifo to write */
-    if (ucs_unlikely((head - ep->fifo_ctl->tail) >= iface->config.fifo_size)) {
-        return UCS_ERR_NO_RESOURCE;
+    /* check if there is room in the remote process's receive FIFO to write */
+    if (!UCT_MM_EP_IS_ABLE_TO_SEND(head, ep->cached_tail, iface->config.fifo_size)) {
+        if (!ucs_arbiter_group_is_empty(&ep->arb_group)) {
```
Comment (shamisp) [inline src/uct/sm/mm/mm_ep.c]: I think it is an "unlike" scenario that we have messages in the arbiter. So I would put unlikely on this condition.

**Comment by shamisp** [comment]
please see my review comments

File: src/uct/sm/mm/mm_ep.c (hunk 3 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: this happens only when we are unable to send, so i'm not sure this is really unlikely

**Comment by shamisp** [comment]
:+1:

**Pull request merged**